### ISSUE TITLE
Displaying a message emits Background Event on Android

### ISSUE THREAD

**Issue by pflenker** [urbanairship/react-native-module#199]
# Preliminary Info

### What Airship dependencies are you using?
3.1.2

### What are the versions of any relevant development tools you are using?

# Report

### What unexpected behavior are you seeing?
When opening a message, all `AppState` listeners fire a `change` event as if the app is going to the background on Android. 

### What is the expected behavior?
Either no `AppState` event is triggered, or there is a way of knowing that a message has been opened via UrbanAirship.

We have code that should only be run when the app transitions from the background to the foreground, but it should not run while the user is using the app, even when he is reading or dismissing a UA message.

For messages opened programatically with `displayMessage`, I can counteract in the code, but messages sent as an in-app notification which open a message on tap, I have no way of knowing that the app is not actually going to the background.

I was assuming that UA fires a `notificationResponse` event in that case, but that seems not to be the case.

### What are the steps to reproduce the unexpected behavior?
Add an AppState listener like this:
```
   AppState.addEventListener('change', (newstate) => console.log(newstate)));
```
Open a message with `displayMessage`, or by interacting with an in app notification which is set to open a message in the message center. Observe that the listener fires on Android.

**Comment by rlepinski** [comment]
React natives app state events are technically wrong. It assumes everything is within the RN activity, but in this case it's being `stopped` but another activity is being started. Not sure how we are going to handle this, the web context might be paused when the activity is stopped preventing your app from receiving events.

**Comment by pflenker** [comment]
If `UrbanAirship.setAutoLaunchDefaultMessageCenter(false)` has been called, you could not display the message directly, but instead emit `showInbox` with the message ID. That way, people could do their own logic before calling ` displayMessage` manually. That has also the benefit that it would be closer to the purpose of the message, as when composing a message the UI says "sends the user to the following message in your message center".

That way, a viable solution exists for anyone who needs to control the app when it goes into background.

**Comment by rlepinski** [comment]
@pflenker, that is how that functionality is supposed to work. Let me take a look

**Comment by pflenker** [comment]
Let me know if you need further info to reproduce that issue!

**Comment by rlepinski** [comment]
@pflenker If I set `setAutoLaunchDefaultMessageCenter` its not showing the message directly, instead its just emitting a showInbox event. However, in my test app notification opening is creating a new activity that is causing the app state change.

Is the autolaunching behavior not working for you?

**Comment by rlepinski** [comment]
I had to set the `android:launchMode="singleInstance"` in order to get react-native to not launch another instance of the main activity, but after that I was able to get the following when setAutoLaunchDefaultMessageCenter was disabled:

```
ReactNativeJS: stateChange: active
ReactNativeJS: showInbox: {"messageId":"tmsmQK8KEemrJgJCv-zO-Q"}
```

**Comment by pflenker** [comment]
I can confirm that `showInbox` is now emitted as expected, so the workaround works! 

Now I will check changing the `launchMode` to get rid of the workaround.

**Comment by pflenker** [comment]
The behaviour with the changed `launchMode` is very weird. It does not open the message center. 
To be clear, here is what I did:
- Added `android:launchMode="singleInstance"` to the main activity
- sent out a message with type "In-App message & Message Center" while the app was active
- Tapped on the In App Message.

Expected behaviour: `showInbox` is emitted, no `AppState` change is emitted.
Actual behaviour: Tapping on the notification creates no visible change, but the debugger reconnects

**Comment by rlepinski** [comment]
Thanks for the info, I will try to reproduce with an IAM

**Comment by rlepinski** [comment]
@pflenker Ok, you want `singleTask`.

In-app message do not emit notification response events. You will only get a showInbox event if its tied to a message.

**Comment by rlepinski** [comment]
Released 3.2.0 with a new `UAMessageView` that you can display an inbox message in directly natively in react. Here is the list of properties - https://github.com/urbanairship/react-native-module/blob/master/js/UAMessageView.js#L57

Going to close the issue for now. If you are still having issues we can reopen.

**Comment by rlepinski** [comment]
We found a bug in the SDK that was probably causing the dual activity launch. Should be fixed in 3.2.1 and 4.0.1
